feat(auth): add auth types

israx opened the pull request:
- Requests Cognito & Category level
- Return Cognito & Category level
- Options Cognito & Category level
- Models Cognito & Category level
- Cognito commands (single auth apis)
- AuthInterface (All apis at the category level)
- AuthClass (singleton with all apis with at cognito level)

israx committed:
feat: add auth types

israx committed:
feat: add indivial auth apis

israx commented on src/auth/auth.ts:
```diff
@@ -0,0 +1,102 @@
+
+  signIn(req: SignInRequest): Promise<AuthSignInResult>;
+
+  confirmSignIn<
```
TBD

israx commented on src/auth/auth.ts:
```diff
@@ -0,0 +1,102 @@
+
+  fetchDevices(): Promise<AuthDevice[]>;
+
+  readonly TOTP: TOTPNamespace;
```
TBD

jimblanc commented on src/auth/auth.ts:
```diff
@@ -0,0 +1,102 @@
+import { UpdateUserAttributeResult } from "./types/result/updateUserAttributeResult";
+import { AuthSignInStep } from "./types/step";
+
+export interface AuthInterface {
```
Mentioned for the API PR as well, but I wonder if we should declare functions directly or use a namespace to indicate that we're moving away from these monolithic classes? It looks like we already have this in the cognito commands type defs below, that might be sufficient for the purposes of the RFC.

israx replied:
Thanks Jim for reviewing this PR. Yes I think we'd like to get feedback on API usage, specially on response and error handling. Some APIs offer a `nextStep` that provides information on how to proceed with the authentication process.

jimblanc commented on src/auth/cognito/awsCognitoAuthSession.ts:
```diff
@@ -0,0 +1,31 @@
+    userSub: Result<string>;
+    identityId: Result<string>;
+  };
+  
+  
+  
+  
+  export class AWSCognitoAuthSession implements AuthSession {
```
Nit: We should clean up spacing & tabbing before putting this in front of customers. Also had comment on another file about omitting any implementations from the RFC and just focusing on types.

jimblanc commented on src/auth/cognito/cognitoAuth.ts:
```diff
@@ -0,0 +1,185 @@
+} from "./types/options";
+import { TOTPNamespace } from "./types/totpNameSpace";
+
+export class AuthClass implements AuthInterface {
```
Can we omit any implementation files (even if they have empty implementations) from the RFC and just focus on the type definitions?

jimblanc commented on src/auth/cognito/commands/index.ts:
```diff
@@ -0,0 +1,162 @@
+} from "../types/options";
+import { TOTPNamespace } from "../types/totpNameSpace";
+
+declare function signUp(
```
This file looks good! I think this is kind of how we want to structure the RFC.

jimblanc commented on src/auth/cognito/commands/index.ts:
```diff
@@ -0,0 +1,162 @@
+
+const TOTP: TOTPNamespace = {
+  setup: function (): Promise<SetupSoftwareTokenResult> {
+    throw new Error("Function not implemented.");
```
Could we omit implementation from these TOTP functions?

israx replied:
done

jimblanc commented on src/auth/cognito/types/credentials.ts:
```diff
@@ -0,0 +1,34 @@
+ */
+export type AWSCognitoAuthSession = {
+    isSignedIn: boolean,
+    pluginData: {
```
If all of the fields of `pluginData` are optional, could `pluginData` itself be optional?

jimblanc commented on src/auth/cognito/types/models/AWSAuthDevice.ts:
```diff
@@ -0,0 +1,9 @@
+import { AuthAttributeType } from "../../../types/models/authAttributeType";
+
+export interface AWSAuthDevice extends AuthDevice {
+  attributes: AuthAttributeType;
+  createDate?: Date;
+  lastAuthenticatedDate?: Date;
+  lastModifiedDate?: Date;
```
Nit: In consistent tabbing with other files

jimblanc commented on src/auth/cognito/types/models/validationData.ts:
```diff
@@ -0,0 +1 @@
+export type ValidationData = { [key: string]: string }
```
Nit: Should make sure each file has a new line at the end

jimblanc commented on src/auth/cognito/types/options/index.ts:
```diff
@@ -0,0 +1,13 @@
+export {AutoSignInOptions} from "./autoSignInOptions"
```
For the purposes of the RFC, I wonder if it would make sense to merge all of the option types into a single file to simplify review for customers? Regardless of what we do for the final implementation. Thoughts? (same argument could be applied to requests, results, steps, etc.)

israx replied:
yes, that's a good feedback. I just updated it. Now there are 15 files instead of 91

jimblanc commented on src/auth/cognito/types/result/signOutResult.ts:
```diff
@@ -0,0 +1,43 @@
+import { SignOutResult } from "../../../types/result";
+
+export class CompleteSignOut implements SignOutResult {
```
Ditto about removing implementations for the RFC

jimblanc commented on src/auth/types/result/signOutResult.ts:
```diff
@@ -0,0 +1,6 @@
+
```
Empty interface, do we need this type for now?

jimblanc reviewed (commented):
In general looks pretty good! Are there anythings specifically you think we should ask for feedback on for the RFC? I.e. flows that might diverge significantly as far as typing we want specific feedback on? I'm trying to think about how we could boil things down to help customers provide useful feedback.

israx committed:
chore: add any auth attribute

This change avoids the type string consuming elements of the union type

israx committed:
clean up

chore: update totp namespace

chore: change code to confirmationCode on confirmSoftware request

chore: auth error clean up

clean up

clean up

clean up

israx committed:
chore: add getAttributeKey util type

israx committed:
add space between functions
